torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -155,6 +155,139 @@ std::pair<std::string, std::string> codeGeneration(Fusion& fusion) {
+  at::ScalarType arg_data_type = arg.scalar_type();
+  DataType param_data_type = *param->getDataType();
+  bool match = false;
+  switch (arg_data_type) {
+    case at::ScalarType::Int:
+      match = param_data_type == DataType::Int;
+      break;
+    case at::ScalarType::Half:
+      match = param_data_type == DataType::Half;
+      break;
+    case at::ScalarType::Float:
+      match = param_data_type == DataType::Float;
+      break;
+    case at::ScalarType::Bool:
+      match = param_data_type == DataType::Bool;
+      break;
+    default:
+      match = false;
+  }
+  if (!match)
+    msg << "Argument element type is " << arg_data_type
+        << ", but the parameter is " << param_data_type;
+  return match;
+}
+
+bool validateKernelArgScalar(
+    const c10::TypePtr arg_type,
+    const Val* param,
+    std::stringstream& msg) {
+  if (!param->isScalar()) {
+    msg << "Argument is a scalar, but the parameter is not.";
+    return false;
+  }
+  DataType param_type = *param->getDataType();
+  bool match = false;
+  switch (arg_type->kind()) {
+    case c10::TypeKind::IntType:
+      match = param_type == DataType::Int;
+      break;
+    case c10::TypeKind::FloatType:
+      match = param_type == DataType::Float;
+      break;
+    case c10::TypeKind::BoolType:
+      match = param_type == DataType::Bool;
+      break;
+    default:
+      match = false;
+  }
+  if (!match) {
+    msg << "Argument type is " << *arg_type << ", but the parameter is "
+        << param_type;
+  }
+  return match;
+}
+
+bool validateKernelArg(
+    const c10::IValue& arg,
+    const Val* param,
+    int device_index,
+    std::stringstream& msg) {
+  if (arg.type()->kind() != c10::TypeKind::TensorType) {
+    return validateKernelArgScalar(arg.type(), param, msg);
+  } else {
+    return validateKernelArgTensor(arg.toTensor(), param, device_index, msg);
+  }
+}
+
+void validateKernelArgs(
+    const CudaKernel& entry,
+    const at::ArrayRef<IValue>& inputs,
+    const std::vector<at::Tensor>& outputs) {
+  // Check inputs
+  TORCH_INTERNAL_ASSERT(
+      inputs.size() == entry.inputs.size(), "Wrong number of kernel inputs.");
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    const IValue& arg = inputs[i];
+    const Val* param = entry.inputs[i];
+    std::stringstream msg;
+    TORCH_INTERNAL_ASSERT(
+        validateKernelArg(arg, param, entry.device_, msg),
+        "Input argument at position ",
+        i,
+        " is invalid; ",
+        msg.str());
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      entry.outputs.size() != 0,
+      "Kernel should have at least one output tensor.");
+
+  TORCH_INTERNAL_ASSERT(
+      outputs.size() == entry.outputs.size(),
+      "Wrong number of kernel outputs.");
+  for (size_t i = 0; i < outputs.size(); ++i) {
+    const at::Tensor& arg = outputs[i];
+    const Val* param = entry.outputs[i];
+    std::stringstream msg;
+    TORCH_INTERNAL_ASSERT(
+        validateKernelArg(arg, param, entry.device_, msg),
```

Comment:
nitpick: `arg` here is a `at::Tensor` instead of `c10::IValue`. Maybe we want to use `validateKernelArgTensor` directly to avoid constructing an IValue.

Reply:
I didn't notice that. Fixed. Thanks!